language/bytecode-verifier/src/dependencies.rs

```diff
@@ -71,20 +82,48 @@ impl<'a, 'b> Context<'a, 'b> {
             }
             // Module::FuncName -> def handle idx
             for func_def in module.function_defs() {
-                let may_be_called = match func_def.visibility {
-                    Visibility::Public => true,
-                    Visibility::Private => false,
-                };
-                if !may_be_called {
-                    continue;
-                }
                 let func_handle = module.function_handle_at(func_def.function);
                 let func_name = module.identifier_at(func_handle.name);
-                context
-                    .func_id_to_handle_map
-                    .insert((module_id.clone(), func_name.to_owned()), func_def.function);
+                dependency_visibilities.insert(
+                    (module_id.clone(), func_name.to_owned()),
+                    func_def.visibility,
+                );
+                match func_def.visibility {
+                    Visibility::Private => continue,
+                    Visibility::Public | Visibility::Script => {
+                        context
+                            .func_id_to_handle_map
+                            .insert((module_id.clone(), func_name.to_owned()), func_def.function);
+                    }
+                }
+            }
+        }
+
+        for function_def in self_function_defs {
+            let visibility = function_def.visibility;
+            context
+                .function_visibilities
+                .insert(function_def.function, visibility);
+        }
+        for (idx, function_handle) in context.resolver.function_handles().iter().enumerate() {
+            if Some(function_handle.module) == self_module_idx {
+                continue;
             }
+            let owner_module_id = context
+                .resolver
+                .module_id_for_handle(context.resolver.module_handle_at(function_handle.module));
+            let function_name = context.resolver.identifier_at(function_handle.name);
+            let visibility =
+                match dependency_visibilities.get(&(owner_module_id, function_name.to_owned())) {
+                    // This function does not link, so just mark it public
```

Comment:
Maybe a comment here:

The arbitrary visibility set here does not matter because if the function does not link, it will be reported by `verify_imported_functions` anyway